src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs

```diff
@@ -0,0 +1,300 @@
+namespace Particular.TimeoutMigrationTool.NHibernate
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using global::NHibernate;
+    using global::NHibernate.Cfg;
+    using global::NHibernate.Cfg.MappingSchema;
+    using global::NHibernate.Criterion;
+    using global::NHibernate.Mapping.ByCode;
+    using Newtonsoft.Json;
+
+    public class NHibernateTimeoutsSource : ITimeoutsSource
+    {
+        public NHibernateTimeoutsSource(string connectionString, int batchSize, DatabaseDialect databaseDialect)
+        {
+            this.connectionString = connectionString;
+            this.batchSize = batchSize;
+            this.databaseDialect = databaseDialect;
+        }
+
+        private ISessionFactory CreateSessionFactory()
+        {
+            var cfg = new Configuration().DataBaseIntegration(x =>
+                {
+                    x.ConnectionString = connectionString;
+                    x.LogSqlInConsole = true;
+                });
+
+            databaseDialect.ConfigureDriverAndDialect(cfg);
+
+            var mapper = new ModelMapper();
+            mapper.AddMapping<TimeoutEntityMap>();
+            mapper.AddMapping<StagedTimeoutEntityMap>();
+            mapper.AddMapping<MigrationsEntityMap>();
+
+            HbmMapping mapping = mapper.CompileMappingForAllExplicitlyAddedEntities();
+            cfg.AddMapping(mapping);
+            cfg.SetProperty("hbm2ddl.auto", "update");
+
+            return cfg.BuildSessionFactory();
+        }
+
+        public async Task<IToolState> TryLoadOngoingMigration()
+        {
+            using var session = CreateSessionFactory().OpenSession();
+
+            var migration = await session.QueryOver<MigrationsEntity>()
+                .Where(m => m.Status == MigrationStatus.StoragePrepared)
+                .SingleOrDefaultAsync();
+
+            if (migration == null)
+            {
+                return null;
+            }
+
+            async Task<BatchInfo> GetNextBatch()
+            {
+                using var session = CreateSessionFactory().OpenStatelessSession();
+
+                var stagedTimeoutEntities = await session.QueryOver<StagedTimeoutEntity>()
+                    .Where(stagedTimeoutEntity => stagedTimeoutEntity.BatchState != BatchState.Completed)
+                    .SelectList(timeoutEntities => timeoutEntities.SelectGroup(te => te.BatchNumber)
+                        .SelectGroup(te => te.BatchState)
+                        .SelectCount(te => te.Id))
+                    .OrderBy(entity => entity.BatchNumber)
+                    .Asc.ListAsync<object[]>();
+
+                return stagedTimeoutEntities.Select(x => new BatchInfo((int) x[0], (BatchState) x[1], (int) x[2]))
```

Comment:
I was wondering if we could use Nhibernates way of projecting into non-tracked entities here and if that would make it more readable. Not against this approach by the way and maybe it is not worth the extra effort. Just a thought